File: cpp/450.cpp
#include <iostream>

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  TreeNode *deleteNode(TreeNode *root, int key) {
    if (!root)
      return nullptr;
    // searching
    if (key < root->val)
      root->left = deleteNode(root->left, key);
    else if (key > root->val)
      root->right = deleteNode(root->right, key);
    else { // if root is target
      if (!root->left && !root->right) { // root is leaf
        return nullptr;
      } else if (!root->left || !root->right) { // root has one child
        TreeNode *ret = !root->left ? root->right : root->left;
        return ret;
      }
      // if have two child, replace root->val by the largest num in
      // left subtree, and delete it
      TreeNode *n = root->left;
      while (n->right)
        n = n->right;
      root->val = n->val;
      root->left = deleteNode(root->left, n->val);
    }
    return root;
  }
};

File: go/450.go
package main

//Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func deleteNode(root *TreeNode, key int) *TreeNode {
	if root == nil { // root is nil
		return nil
	}
	// look for target
	if root.Val < key {
		root.Right = deleteNode(root.Right, key)
	} else if root.Val > key {
		root.Left = deleteNode(root.Left, key)
	} else { // root is target
		if root.Left == root.Right { // root is leaf
			return nil
		} else if root.Left == nil { // has right child
			return root.Right
		} else if root.Right == nil { // has left child
			return root.Left
		}
		// have two childs, replace root.Val by largest in left subtree
		var n *TreeNode = root.Left
		for n.Right != nil {
		    n = n.Right
		}
		root.Val = n.Val
		root.Left = deleteNode(root.Left, n.Val)
	}
	return root
}
